### PR TITLE
Port conformance tests and automation to remote-signing

### PR DESCRIPTION
**Pull request by qmuntal**
This is almost a clean port from main to remote-signing branch.

This PR also adds `crypto/sha256` and `crypto/sha512` as blank imports to `algorithm.go`. This is because go1.17 does not make them available by default, therefore `Algorithm.newHash` fails for several algorithms.

With this change, CI tests will run on every PR commit.

Fuzzing still don't test roundtripping. Running fuzzing test locally, commenting out the roundtrip test, shows that we still don't achieve it.

### CODE REVIEW

**Comment by qmuntal** [comment]
Coverage test is failing because we went down to ~70% coverage. I would merge this PR anyway, as this is a known issue, and we will work on adding new tests sooner than later.

**Comment by SteveLasker** [comment]
So cool to see these uplifted to the remote-signing branch.
CC: @thomas-fossati, @shizhMSFT, @yogeshbdeshpande

**Review by SteveLasker — APPROVED** [review]
LGTM

File: .github/workflows/ci-go-cover.yml [hunk 1]
```diff
@@ -25,9 +25,11 @@ jobs:
     runs-on: ubuntu-latest
     steps:
     - name: Checkout code
-      uses: actions/checkout@v2
-    - name: Install test cases
-      run: make install
+      uses: actions/checkout@v2
```
Comment (shizhMSFT) [inline .github/workflows/ci-go-cover.yml:28]: should we upgrade to `actions/checkout@v3`?

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -1,25 +1,22 @@
...
-        fetch-depth: 1
-    - name: Get dependencies
-      run: go get -v -t -d ./...
-    - name: Build project
-      run: go build .
-    - name: Install test cases
-      run: make install
+        go-version: ${{ matrix.go-version }}
+    - name: Checkout code
+      uses: actions/checkout@v2
```
Comment (shizhMSFT) [inline .github/workflows/ci.yml:20]: same here

Comment (shizhMSFT) [inline .github/workflows/ci.yml:20]: same here

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -1,25 +1,22 @@
...
+on: [push, pull_request]
 jobs:
 
   # Test on various OS with default Go version. 
   tests:
-    name: Test on ${{matrix.os}}
+    strategy:
+      fail-fast: false
+      matrix:
+        go-version: [1.17.x, 1.18.x]
```
Comment (shizhMSFT) [inline .github/workflows/ci.yml:12]: Do we still need `1.17.x` since `go.mod` is now `1.18`?

Comment (shizhMSFT) [inline .github/workflows/ci.yml:12]: Do we still need `1.17.x` since `go.mod` is now `1.18`?

File: algorithm.go [hunk 4]
```diff
@@ -2,6 +2,8 @@ package cose
 
 import (
 	"crypto"
+	_ "crypto/sha256"
+	_ "crypto/sha512"
```
Comment (shizhMSFT) [inline algorithm.go:6]: Curious why we don't need `crypto/sha384`. It must be imported somewhere.

Comment (shizhMSFT) [inline algorithm.go:6]: Can you document those import in the package documentation since it might not be desired?

File: conformance_test.go [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+var encMode, _ = cbor.CanonicalEncOptions().EncMode()
+
+func decodeHeaders(protected, unprotected []byte) (hdr cose.Headers, err error) {
+	// test-vectors encodes the protected header as a map instead of a map wrapped in a bstr.
+	// UnmarshalFromRaw expects the former, so wrap the map here before passing it to UnmarshalFromRaw.
+	hdr.RawProtected, err = encMode.Marshal(protected)
+	if err != nil {
+		return
+	}
+	hdr.Protected = make(cose.ProtectedHeader)
```
Comment (shizhMSFT) [inline conformance_test.go:245]: It is not necessary to do initialization. I think `UnmarshalFromRaw()` will handle it.

Comment (shizhMSFT) [inline conformance_test.go:245]: It is not necessary to do initialization. I think `UnmarsahlFromRaw()` will handle it.

File: fuzz_test.go [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+			return
+		}
+		_, err := msg.MarshalCBOR()
+		if err != nil {
+			t.Fatalf("failed to marshal valid message: %s", err)
+		}
+		// TODO: uncomment the roundtrip test once https://github.com/veraison/go-cose/pull/17 lands.
+		// if !bytes.Equal(b, got) {
+		// 	t.Fatalf("roundtripped message has changed, got: %v, want: %v", got, b)
+		// }
```
Comment (shizhMSFT) [inline fuzz_test.go:56]: https://github.com/veraison/go-cose/pull/17 landed ::rocket::

Comment (shizhMSFT) [inline fuzz_test.go:56]: https://github.com/veraison/go-cose/pull/17 landed ::rocket::

File: .github/workflows/ci-go-cover.yml (hunk 1 above)
Comment (qmuntal) [reply .github/workflows/ci-go-cover.yml:28]: Sure

File: .github/workflows/ci.yml (hunk 2 above)
Comment (qmuntal) [reply .github/workflows/ci.yml:20]: Sure x2

File: .github/workflows/ci.yml (hunk 3 above)
Comment (qmuntal) [reply .github/workflows/ci.yml:12]: We still want to test against older versions, so we are sure those are supported.

The `go` directive in `go.mod` does not set the minimum supported Go version, one can still use older versions. It indicates to the compiler which Go features are allowed, i.e. `go 1.17` would mean fuzzing and generics are not available even if go1.18 compiler is used.

The `go` directive does help on keeping backwards compatibility with older versions, but there are other ways to achieve it. For example, our fuzz tests are placed in `fuzz_test.go`, which is gated behind the build constraint `//go:build go1.18`. Therefore go1.17 and older compilers will be able to test the module, but without running the fuzz tests.

File: algorithm.go (hunk 4 above)
Comment (qmuntal) [reply algorithm.go:6]: You mean removing the imports, adding them to `conformance_test.go` so tests can pass and documenting this requirement?

File: conformance_test.go (hunk 5 above)
Comment (qmuntal) [reply conformance_test.go:245]: True!

File: algorithm.go (hunk 4 above)
Comment (qmuntal) [reply algorithm.go:6]: There is no such a package. `sha384` is imported as part of `crypto/sha512`.

[the remark this answers is not in the document]
Comment (qmuntal) [reply algorithm.go:6]: There is no such a package. `sha384` is imported as part of `crypto/sha512`.

**Review by shizhMSFT — APPROVED** [review]
LGTM

**Pull request merged**